lightning/src/chain/channelmonitor.rs

```diff
@@ -1875,17 +1839,43 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 		Vec::new()
 	}
 
-	/// Called by SimpleManyChannelMonitor::block_connected, which implements
-	/// ChainListener::block_connected.
-	/// Eventually this should be pub and, roughly, implement ChainListener, however this requires
-	/// &mut self, as well as returns new spendable outputs and outpoints to watch for spending of
-	/// on-chain.
-	fn block_connected<B: Deref, F: Deref, L: Deref>(&mut self, txn_matched: &[&Transaction], height: u32, block_hash: &BlockHash, broadcaster: B, fee_estimator: F, logger: L)-> Vec<(Txid, Vec<TxOut>)>
+	/// Checks if a given transaction matches the funding transaction or spends any watched outputs.
+	fn does_tx_match(&self, tx: &Transaction) -> bool {
+		let funding_txo = self.get_funding_txo();
+		if tx.txid() == funding_txo.0.txid {
+			for output in tx.output.iter() {
+				if output.script_pubkey == funding_txo.1 {
+					return true;
+				}
+			}
+		}
+
+		for input in tx.input.iter() {
+			for (txid, outputs) in self.get_outputs_to_watch().iter() {
+				for (idx, _script_pubkey) in outputs.iter().enumerate() {
+					if txid == &input.previous_output.txid && idx == input.previous_output.vout as usize {
+						return true;
+					}
+				}
+			}
+		}
+
+		false
+	}
+
+	/// Determines if any HTLCs have been resolved on chain in the connected block.
+	///
+	/// Returns any new outputs to watch from `txdata`. After called, these are also included in
+	/// [`get_outputs_to_watch`].
```

Comment:
This definitely needs a discussion of the requirements for rescan (that you have to do so before connecting the next block, that it is ok to pass channel-provided updates to the monitor in between), and how you can avoid them (that you need not care whatsoever if you provide the full block data, which should be default if you have it, or if you provide the full set of txn that were in the block which spent any previously-matched txn).

Reply:
> Can the rescan be avoided given the block is immediately filtered at the start of this function?

Oof, right, that sounds like a bug on its own (we should do the trick of "if it spends something we already matched, it also matches", I think, not sure if there's anything more efficient) :).

> Here, by rescan, I assume you mean call block_connected again with the same block (possibly re-fetched if you don't have full txdata).

Yes, that you (should be able to) avoid calling it again with the block's data refreshed if you provide full block data (or match anything which spends a previous match).

> Not sure if I'm parsing this correctly. Could you confirm my understanding?

ie "User filters block with the match criteria broadened to 'anything that matches or anything which spends a transaction in the same block which matched'".

Reply:
> Oof, right, that sounds like a bug on its own (we should do the trick of "if it spends something we already matched, it also matches", I think, not sure if there's anything more efficient) :).

@TheBlueMatt I attempted implementing this but am getting some test failures. Here's the branch:

https://github.com/jkczyz/rust-lightning/tree/2020-06-refactor-chain-listener-lazy-filter

Specifically, see: https://github.com/jkczyz/rust-lightning/commit/c4aad5b7789c533c31b95d819bd72cb354190360.

Does this look like the right approach? Could the failures be related to the problem that @ariard noted earlier: https://github.com/rust-bitcoin/rust-lightning/pull/649#issuecomment-683500920? Or is it worth investigating whether the tests might be fragile?

Reply:
Hmm, I was thinking more like this, though it also fails tests, but seemingly fewer tests fail. May be worth playing with that, but I think in either case its likely the tests just being too specific.

```
@@ -1878,7 +1878,20 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
                      F::Target: FeeEstimator,
                                        L::Target: Logger,
        {
-               let txn_matched = &txdata.iter().filter(|&&(_, tx)| self.does_tx_match(tx)).map(|e| *e).collect::<Vec<_>>();
+               let mut match_set = HashSet::new();
+               let txn_matched = &txdata.iter().filter(|&&(_, tx)| {
+                       let mut matches = self.does_tx_match(tx);
+                       for inp in tx.input.iter() {
+                               if matches { break; }
+                               if match_set.contains(&inp.previous_output.txid) {
+                                       matches = true;
+                               }
+                       }
+                       if matches {
+                               match_set.insert(tx.txid());
+                       }
+                       matches
+               }).map(|e| *e).collect::<Vec<_>>();
                for &(_, tx) in txn_matched {
                        let mut output_val = 0;
                        for out in tx.output.iter() {
```

Reply:
@TheBlueMatt The failing tests with your solution are the same tests that failed before I added the filter step. I tried adding the test fixes given by @ariard in https://github.com/ariard/rust-lightning/commit/b893cbfdd88391208dca62ab1fe872a7406ff661, and they do indeed result in the tests passing.

However, I'm not sure I understand why this is considered a viable fix. It breaks the transactions across two blocks, but what would prevent a peer from getting the transactions into a single block?

Reply:
Huh, right, yea, I definitely agree that that's a bogus fix - that seems more like adapting the test to the behavior rather than testing.